### ISSUE TITLE
Do we want to maintain a `napari-io` plugin like this?

### ISSUE THREAD

**Issue by sofroniewn** [sofroniewn/napari-io-test#1]
@tlambert03 - following up on discussion in https://github.com/napari/napari/pull/1023#issuecomment-599152375 I used [cookiecutter-napari-plugin](https://github.com/napari/cookiecutter-napari-plugin) to make this plugin.

It has a folder called `readers` which can contain files such as `imageio_reader.py`, `npy_reader.py`, and eventually could contain `mat_reader.py`, `pims_reader.py`, `meshio_reader.py`, `shapely_reader.py`, `geojson_reader.py`, `czi_reader.py`, `dv_reader.py`, `nd2_reader.py`, `bioformats_reader.py` etc. all of which are very minimal wrappers / converters from underlying libraries into the format napari expects.

I think the advantage of this library is the curation and testing that we could do to ensure that files are able to be read (i'm imagining we in parallel start curating a dataset of files that could be installed and tested by our CI). Another advantage is that for napari users it becomes very simple to just install one thing and have things work, rather than having to chase down and install many things.

Note here that we're able to unify across readers for different data types, shapes, images, meshes etc. A downside is that this package could have so many many dependencies that things breaking / install problems might be common. We'd have to think about what our attitude towards submissions from others would be, we could be very liberal or conservative, especially if we though about other domains of science where file types are very different (physics, astronomy, geospatial etc.)

Alternatively we try and get napari hook impls added to the various underlying libraries directly, or we have lots of `napari-io-*` libraries that do the translation. Or we try and pick a smaller number of libraries to drive standardization too (i.e. get the czi reader into imageio). That still won't help with the mesh and shape reading though.

If we do do this should it live at `napari/napari-io` which in some senses gives it an "official branding" or should it live somewhere else?

My current opinion is to go for it at `napari/napari-io`, set up rigorous testing / CI and try and be as comprehensive as possible, across domains, file types, and napari layer types.

Curious what your vision for all this is @tlambert03? Also curious what @jni and @royerloic thinks.

**Comment by tlambert03** [comment]
I like it!  I had been thinking after the recent request for `.mat` support that relatively easy stuff like that might be best done in a curated "monorepo" like this... before we start seeing a lot of little packages and it all becomes confusing.  I agree fully with your assessment that we could do a lot here with curation and testing, and it would also be useful to have one place to talk about navigating dependency conflicts and "priority" for readers.

Of course, I'm still glad that anyone will be able to write and use their own reader plugins, but I think the average user would probably benefit a lot from a big `napari-io` package, and contributors to the package could, in turn, try to tune its performance for a broad range of file inputs.

anyway I'm supportive of the `napari/napari-io` approach, (at least until someone convinces me otherwise 😄)!

**Comment by sofroniewn** [comment]
That's great @tlambert03 - if we like this idea, then I think there are also significant advantages to having this repo ready to be pip installed with the `0.3.0` release and being able to immediately point people towards it who are interested in providing file-io readers and encourage contributions here from the start.

**Comment by royerloic** [comment]
I like it.
Some thoughts: I would still want of course to be able to open Tiff, Zarr, and PNG, JPG, and other _very_ standard files directly without installing an additional package. But, there are lots of other (exotic) file formats that we might want to see in napari-io that would drag potentially contentious dependencies etc..  So I like the idea of decoupling the two types (standard, from less so).

**Comment by sofroniewn** [comment]
> I would still want of course to be able to open Tiff, Zarr, and PNG, JPG, and other very standard files directly without installing an additional package

Agreed, I don't think we would loose the basic io we provide now, the plan would only be to provide many additional through `napari-io`

**Comment by jni** [comment]
I'm unsure about this. I think it would be cleaner to have smaller packages, together with a metapackage that collects all. There is nothing stopping us from creating `napari-io` (I think you already did that, @sofroniewn?) that depends on a bunch of smaller plugins. We could even have sub-packages, `pip install napari-io[geojson]`. Conda isn't there yet but I don't think they are far off, and it can be simulated by small packages + dependencies.

Having said this, it does create overhead: harder to manage a large set of repos. So, I'm happy to go with the monorepo for now, if you all prefer. Just wanted to suggest the subpackage alternative for consideration.

**Comment by sofroniewn** [comment]
> I'm unsure about this. I think it would be cleaner to have smaller packages, together with a metapackage that collects all. There is nothing stopping us from creating napari-io (I think you already did that, @sofroniewn?) that depends on a bunch of smaller plugins.

This more modular approach might give us lots more flexibility, making testing easier, faster etc. It might be a good idea. The difficulty I see is managing all the repos. But with more granular repos - it might be easier to grant more people read / write access to repos to help with the maintenance burden and github notifications are pretty good if you subscribe to things.

One option then might be to create a brand new GitHub organization `napari-io` that inside it has one main repo `napari-io` that is a simple meta package which has a list of other packages all of which are located at `napari-io/napari-io-nd2` or `napari-io/napari-io-czi` etc.

Having the new GitHub Org might make organization much simpler and less error prone then putting them inside the main `napari` repo as there could be 100s of these reader plugins. I kinda like this idea. We'd still have to think about testing (see #4) - is there one `binaries` repo with raw data in it that the other packages can now pull from, or does each package have their own test data.

Once we flesh out the list of potential packages / file types in #2 it might be easier. Thoughts @tlambert03 @jni? Also @jni any idea what others in the community who work on file-io packages might prefer / find more natural.

**Comment by jni** [comment]
@sofroniewn your idea of distributed management has pushed me a little over the edge towards the meta-repo. This repo could then be a curated list of not just our own high-quality plugins, but community-based one. e.g. if the bioformats folks decide to maintain one, we could add it. If the Zeiss folks create an official one (they seem to like napari! :smile:), we could add it. Etc. It wouldn't have to live in our org.

In that spirit, I think possibly we should keep the repos under the napari org, as maybe most plugins will be external. I think orgs scale pretty well to hundreds, and even low thousands, of repos.

**Comment by sofroniewn** [comment]
I like how you're thinking too @jni  - but have some questions about the details below

> This repo could then be a curated list of not just our own high-quality plugins, but community-based one. 

@tlambert03 how do you think this meta repo (let's call it `napari-io`) would look in practice? The individual repos would all conform to the standard hook-spec and discovery mechanisms. Would this meta-repo just be a list with all those repos so that `pip install napari-io` installs all the individual repos? Seems like from the perspective of https://github.com/napari/napari/pull/1023 and sorting plugins it would be nice to still allow all those individual repos to be sortable. Would the meta repo do anything else? Aggregate documentation? Make decisions about which `tif` reader to prioritize (for example)? Aggregate testing / CI?

Do we even need the meta repo if we're going to maintain distributions of napari, see https://github.com/napari/napari/issues/1001#issuecomment-591586740, and have search and discovery built into napari (i.e. type in ND2 in a search bar and `napari-io-nd2` pops up?). Maybe instead we just thing about what it means to maintain and test a distribution (i.e. pip install all repos, run the tests of the different repos, run some integration tests)?

> In that spirit, I think possibly we should keep the repos under the napari org, as maybe most plugins will be external. 

I didn't quite follow this sentence @jni - are you suggesting that for the individual repos we should promote:
1. `napari/napari-io-nd2`, `napari/napari-io-czi`, `napari/napari-io-meshio`, etc.
2. `napari-io/napari-io-nd2`, `napari-io/napari-io-czi`, `napari-io/napari-io-meshio`, etc.
3. `nikon/napari-io-nd2`, `zeiss/napari-io-czi`, etc. `meshio/napari-io-meshio`
4. `nikon/nd2-reader`, `zeiss/czi-reader`, etc. `meshio/meshio`

The meta repo could either live at:
A. `napari/napari-io`
B. `napari-io/napari-io`

> I think orgs scale pretty well to hundreds, and even low thousands, of repos.

Does this mean you think adding a `napari-io` org is not needed, and it will be clear what's going on if we have `napari/napari-io-nd2`, `napari/napari-io-czi`, etc.? I think that is fair. Might be easier for us to just maintain one org. My only thoughts were around permissions / org memberships. I'd want to look into what it is like to give someone full permissions on just one repository within an org so that say a nikon fan could commit to `napari/napari-io-nd2` and a zeiss fan could commit to `napari/napari-io-czi`. If we feel confident with all that then just having one org is fine, if we're worried about that then having a separate `napari-io` org might be nice. Also as we make more plugins would we ever want a `napari-plugins` org or would be put things in the main org or would we never try and centralize?

Lots of questions here, curious what @jni @tlambert03 and others think

**Comment by tlambert03** [comment]
Might be getting to the point where a conversation is easier.  a few scattered thoughts:

- if the ultimate goal is to have external devs maintaining plugins for specific file-types (like the "zeiss fan" and "nikon fan" in your example), then I don't think it should be in our org at all.  That's a true plugin.
- if we are _not_ creating our own `napari_get_reader` implementation with logic that actively decides which reader function to use under certain cases, then I'm not sure I see the need for a metapackage that actually installs everything, as opposed to a pure "listing" type repo (ala [awesome-python](https://github.com/vinta/awesome-python)), where people can pick and choose what they need for their applications.
- if we _do_ think that we could facilitate a large number of use cases by combining a small number of external dependencies (e.g. pims, imageio, scipy.io, etc...) under one `napari_get_reader` function, then I think it could work to have `napari-io` be a "good first io plugin to install" that doesn't preclude installing additional third-party plugins.  And in that case, `napari-io` would still be under napari/org, and would accept PRs as usual.

**Comment by sofroniewn** [comment]
@danielballan has asked 
> Could a SciPy-wide Reader protocol work with napari?

one the main napari repo - https://github.com/napari/napari/issues/1066.

I plan on responding to @danielballan in detail there don't want to split the conversation into too many places, but I'm not sure if @danielballan has seen this thread (and other issues on this toy repo), but it might give him some insight into preliminary ideas napari has been tossing around so far.

I will briefly add I love the idea of the of standardization lazy file loading across the python ecosystem, and that napari doesn't want to lock anyone into any walled gardens, but would love to be part of these efforts.

**Comment by danielballan** [comment]
I hadn’t found this issue yet. Thanks for tagging me here.

**Comment by danielballan** [comment]
I'll take @sofroniewn's invitation to comment here, humbling hoping that I am not wading into this discussion like the Kool-Aid Man walking through a wall.

@sofroniewn:

> This more modular approach might give us lots more flexibility, making testing easier, faster etc. It might be a good idea. The difficulty I see is managing all the repos. But with more granular repos - it might be easier to grant more people read / write access to repos to help with the maintenance burden and github notifications are pretty good if you subscribe to things.

Intake, which I have started contributing to over the last year or so, uses the "many small repos" approach. As I just rewrote their plugin discovery system to use `entry_points` instead of scraping `sys.path` to packages that are named `intake*`, I bring recent [first-hand experience with the joy of applying a tedious update to many small repos](https://github.com/intake/intake/issues/392). :-D

_Even so_ I think many-small-repos is the right way to go. The points mentioned above are right on. To expand a bit:
- Separate issue trackers mean that the proverbial Nikon Fan and Zeiss Fan don't have to receive GH notifications from code they don't use.
- I/O dependencies can be fiddly in CI, and keeping them separate prevents one problematic dependency from gumming up development on everything else.
- It puts the centrally-maintained plugins on closer to the same footing as ones that happen to be maintained in external GH repos.
- It can make it easier to tell which plugins are important to user community and should get more resources.

@tlambert03:

> if we are not creating our own napari_get_reader implementation with logic that actively decides which reader function to use under certain cases, then I'm not sure I see the need for a metapackage that actually installs everything, as opposed to a pure "listing" type repo (ala awesome-python), where people can pick and choose what they need for their applications.

This seems right to me as well. For [a work project of mine that has file _writing_ plugins](https://blueskyproject.io/suitcase/) we thought we would need a meta-package repository with "commonly useful formats," but a year or two in now we still haven't needed one.

It seems you would only _need_ one here if you want to explicitly list precedence order of readers in the case where there are multiple compatible options.

**Comment by tlambert03** [comment]
thanks for chiming in @danielballan!  your input is super helpful.  I'm actually reading your reader protocol blog post at the moment (with great interest) and will try to share some thoughts over in https://github.com/napari/napari/issues/1066 sometime today.  Definitely sounds like there's a great deal of overlap in interest/philosophy/application

**Comment by sofroniewn** [comment]
> I think many-small-repos is the right way to go. The points mentioned above are right on. To expand a bit:
>
> - Separate issue trackers mean that the proverbial Nikon Fan and Zeiss Fan don't have to receive GH notifications from code they don't use.
> - I/O dependencies can be fiddly in CI, and keeping them separate prevents one problematic dependency from gumming up development on everything else.
> - It puts the centrally-maintained plugins on closer to the same footing as ones that happen to be maintained in external GH repos.
> - It can make it easier to tell which plugins are important to user community and should get more resources.

Yeah, those are great reasons.

> we thought we would need a meta-package repository with "commonly useful formats," but a year or two in now we still haven't needed one.

Yeah, I think our thoughts might be trending in that direction.

Thanks for theses additions!!! I think we will be philosophically aligned, and then we can get down to the details :-)
